sbin/mount_lfs/mount_lfs.8
.\"	BSDI mount_lfs.8,v 2.1 1995/02/03 07:23:35 polk Exp
.\"
.\" Copyright (c) 1993, 1994
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)mount_lfs.8	8.5 (Berkeley) 3/30/94
.\"
.Dd "March 30, 1994"
.Dt MOUNT_LFS 8
.Os BSD 4.4
.Sh NAME
.Nm mount_lfs
.Nd mount a log-structured file system
.Sh SYNOPSIS
.Nm mount_lfs
.Op Fl dns
.Op Fl o Ar options
.Ar special
.Ar node
.Sh DESCRIPTION
The
.Nm mount_lfs
command attaches a log-structured file system
.Ar special
device on to the file system tree at the point
.Ar node .
In addition, the
.Xr lfs_cleanerd 8
utility is invoked to clean the file system periodically.
.Pp
This command is normally executed by
.Xr mount 8
at boot time.
.Pp 
Options can be specified as a single letter flag,
with an optional value, in the standard
.Xr getopt 3
style. 
Alternatively, options can be given as arguments via the
.Fl o
flag.
The long forms below appear in brackets.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl d Op Nm cleanerdebug
Run
.Xr lfs_cleanerd 8
in debug mode.
.It Fl o
Options are specified with a
.Fl o
flag followed by a comma separated string of options.
See the
.Xr mount 8
man page for possible options and their meanings.
.It Fl n Op Nm noclean
Don't start
.Xr lfs_cleanerd 8
on the file system.
.It Fl s Op Nm shortrds
Cause
.Xr lfs_cleanerd 8
to read data in small chunks when cleaning the file system.
.El
.Sh SEE ALSO
.Xr mount 2 ,
.Xr unmount 2 ,
.Xr fstab 5 ,
.Xr lfs_cleanerd 8 ,
.Xr mount 8
.sp
.Rs
.%A Ousterhout and Douglis
.%D 1989
.%T "Beating the I/O Bottleneck: A Case for Log-structured File Systems"
.%J "Operating Systems Review"
.%V Vol. 23
.%N No. 1
.%P pp. 11-27
.%O "also available as Technical Report UCB/CSD 88/467"
.Re
.Rs
.%A Rosenblum and Ousterhout
.%D 1991
.%T "The Design and Implementation of a Log-Structured File System"
.%J "ACM SIGOPS Operating Systems Review"
.%V Vol. 25
.%N No. 5
.Re
.Rs
.%A Seltzer
.%D 1992
.%T "File System Performance and Transaction Support"
.%B "PhD Thesis, University of California, Berkeley"
.%O "also available as Technical Report UCB/ERL M92"
.Re
.Rs
.%A Seltzer, Bostic, McKusick and Staelin
.%D 1993
.%T "An Implementation of a Log-Structured File System for UNIX"
.%J "Proc. of the Winter 1993 USENIX Conf."
.%P pp. 315-331
.Re
.Sh HISTORY
The
.Nm mount_lfs
function first appeared in 4.4BSD.


sbin/mount_lfs/mount_lfs.c
/*	BSDI mount_lfs.c,v 2.1 1995/02/03 07:23:39 polk Exp	*/

/*-
 * Copyright (c) 1993, 1994
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
static char copyright[] =
"@(#) Copyright (c) 1993, 1994\n\
	The Regents of the University of California.  All rights reserved.\n";
#endif /* not lint */

#ifndef lint
static char sccsid[] = "@(#)mount_lfs.c	8.3 (Berkeley) 3/27/94";
#endif /* not lint */

#include <sys/param.h>
#include <sys/mount.h>

#include <err.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "mntopts.h"
#include "pathnames.h"

int short_rds, cleaner_debug, noclean;

struct mntopt mopts[] = {
	MOPT_STDOPTS,
	MOPT_UPDATE,
	{ "clean", 1, 1, 0, &noclean, },
	{ "cleanerdebug", 0, 1, 0, &cleaner_debug, },
	{ "shortrds", 0, 1, 0, &short_rds, },
	{ NULL }
};

void	usage __P((void));
void	invoke_cleaner __P((char *));


int
main(argc, argv)
	int argc;
	char *argv[];
{
	struct ufs_args args;
	int ch, mntflags;
	char *fs_name, *options;

	options = NULL;
	mntflags = noclean = 0;
    	memset(&args, 0, sizeof(args));

	while ((ch = getopt(argc, argv, "dno:s")) != EOF)
		switch (ch) {
		case 'd':
			cleaner_debug = 1;
			break;
		case 'n':
			noclean = 1;
			break;
		case 'o':
			getmntopts(optarg, mopts, &mntflags);
			break;
		case 's':
			short_rds = 1;
			break;
		case '?':
		default:
			usage();
		}
	argc -= optind;
	argv += optind;

	if (argc != 2)
		usage();

        args.fspec = argv[0];	/* the name of the device file */
	fs_name = argv[1];	/* the mount point */

	if (mntflags & MNT_RDONLY)
		args.export.ex_flags = MNT_EXRDONLY;
	else
		args.export.ex_flags = 0;

	if (mount(MOUNT_LFS, fs_name, mntflags, &args))
		if (errno == ENODEV)
			errx(1, "%s on %s: filesystem type not configured",
				args.fspec, fs_name);
		else
			err(1, "%s on %s", args.fspec, fs_name);

	if (!noclean)
		invoke_cleaner(fs_name);
		/* NOTREACHED */

	exit(0);
}

void
invoke_cleaner(name)
	char *name;
{
	char *args[6], **ap = args;

	/* Build the argument list. */
	*ap++ = _PATH_LFS_CLEANERD;
	if (short_rds)
		*ap++ = "-s";
	if (cleaner_debug)
		*ap++ = "-d";
	*ap++ = name;
	*ap = NULL;

	execv(args[0], args);
	err(1, "exec %s", _PATH_LFS_CLEANERD);
}

void
usage()
{
	(void)fprintf(stderr,
		"usage: mount_lfs [-dns] [-o options] special node\n");
	exit(1);
}
